[WIP] Check hosts for upcoming expiring certificates

tbielawa opened the pull request:
Identifies openshift related certificates on remote hosts, unique's them (some are in multiple locations on a single host), find their expiration dates, and then generates output per certificate that will expire within a defined expiration window (now + **X** days).

* This feature comes from https://trello.com/c/OpWOZ3rx/124-3-warn-admins-that-their-certificates-will-soon-expire

# Customizing

The expiration warning parameter can be set to your choice. For example, during testing you may want to set the "warning window" to a large number, like nearly 2 years in days (730). 

* In ``playbooks/byo/check-cert-expiry.yaml`` set ``cert_expiry_days_warning`` to the warning window. Certificates that are 'expiring' have an expiration date ``now < cert-expiration date <= now+warning-window days``

* In ``playbooks/byo/check-cert-expiry.yaml`` set ``hosts`` to whatever hostgroup you want to check

# Usage

Ideal situation, nothing is expiring soon:

```
$ ansible-playbook -v -i ./inventory/byo/hosts.tbielawa.cert-expiry-multi playbooks/adhoc/check-cert-expiry.yaml
Using /etc/ansible/ansible.cfg as config file

PLAY [all] *********************************************************************

TASK [find] ********************************************************************
Thursday 21 July 2016  11:36:19 -0400 (0:00:00.026)       0:00:00.026 *********

...

TASK [debug] *******************************************************************
Thursday 21 July 2016  11:37:37 -0400 (0:00:00.092)       0:00:10.466 ********* 
ok: [54.204.157.7] => {
    "msg": "Certificate expiration warning window is 30 days"
}

TASK [fail] ********************************************************************
Thursday 21 July 2016  11:37:37 -0400 (0:00:00.056)       0:00:10.523 ********* 

PLAY RECAP *********************************************************************
54.204.157.7               : ok=6    changed=0    unreachable=0    failed=0   
54.88.146.77               : ok=5    changed=0    unreachable=0    failed=0   
```

Assuming an unrealistically high warning window (730 days), we could run the playbook and get the warnings we're looking for

```
TASK [debug] *******************************************************************
Thursday 21 July 2016  11:39:11 -0400 (0:00:00.093)       0:00:11.935 *********
ok: [54.204.157.7] => {
    "msg": "Certificate expiration warning window is 730 days"
}
  
TASK [fail] ********************************************************************
Thursday 21 July 2016  11:39:11 -0400 (0:00:00.052)       0:00:11.988 *********
failed: [54.204.157.7] (item={'cert': u'/etc/origin/node/system:node:ec2-107-23-232-25.compute-1.amazonaws.com.crt', 'expiry': u'Jul 19 21:25:46 2018 GMT', 'days_remaining': 728, 'subject': u'O=system:nodes, CN=sy
stem:node:ec2-107-23-232-25.compute-1.amazonaws.com'}) => {"failed": true, "item": {"cert": "/etc/origin/node/system:node:ec2-107-23-232-25.compute-1.amazonaws.com.crt", "days_remaining": 728, "expiry": "Jul 19 21
:25:46 2018 GMT", "subject": "O=system:nodes, CN=system:node:ec2-107-23-232-25.compute-1.amazonaws.com"}, "msg": "On host '54.204.157.7' the cert /etc/origin/node/system:node:ec2-107-23-232-25.compute-1.amazonaws.
com.crt with CommonName O=system:nodes, CN=system:node:ec2-107-23-232-25.compute-1.amazonaws.com expires in 728 days (Jul 19 21:25:46 2018 GMT)"}
failed: [54.88.146.77] (item={'cert': u'/etc/origin/generated-configs/node-ec2-107-23-232-25.compute-1.amazonaws.com/system:node:ec2-107-23-232-25.compute-1.amazonaws.com.crt', 'expiry': u'Jul 19 21:25:46 2018 GMT
', 'days_remaining': 728, 'subject': u'O=system:nodes, CN=system:node:ec2-107-23-232-25.compute-1.amazonaws.com'}) => {"failed": true, "item": {"cert": "/etc/origin/generated-configs/node-ec2-107-23-232-25.compute
-1.amazonaws.com/system:node:ec2-107-23-232-25.compute-1.amazonaws.com.crt", "days_remaining": 728, "expiry": "Jul 19 21:25:46 2018 GMT", "subject": "O=system:nodes, CN=system:node:ec2-107-23-232-25.compute-1.amaz
onaws.com"}, "msg": "On host '54.88.146.77' the cert /etc/origin/generated-configs/node-ec2-107-23-232-25.compute-1.amazonaws.com/system:node:ec2-107-23-232-25.compute-1.amazonaws.com.crt with CommonName O=system:
nodes, CN=system:node:ec2-107-23-232-25.compute-1.amazonaws.com expires in 728 days (Jul 19 21:25:46 2018 GMT)"}
failed: [54.204.157.7] (item={'cert': u'/etc/origin/node/server.crt', 'expiry': u'Jul 19 21:25:52 2018 GMT', 'days_remaining': 728, 'subject': u'CN=107.23.232.25'}) => {"failed": true, "item": {"cert": "/etc/origi
n/node/server.crt", "days_remaining": 728, "expiry": "Jul 19 21:25:52 2018 GMT", "subject": "CN=107.23.232.25"}, "msg": "On host '54.204.157.7' the cert /etc/origin/node/server.crt with CommonName CN=107.23.232.25
 expires in 728 days (Jul 19 21:25:52 2018 GMT)"}
failed: [54.88.146.77] (item={'cert': u'/etc/origin/master/master.etcd-client.crt', 'expiry': u'Jul 19 21:06:09 2017 GMT', 'days_remaining': 363, 'subject': u'CN=ec2-54-209-71-248.compute-1.amazonaws.com'}) => {"f
ailed": true, "item": {"cert": "/etc/origin/master/master.etcd-client.crt", "days_remaining": 363, "expiry": "Jul 19 21:06:09 2017 GMT", "subject": "CN=ec2-54-209-71-248.compute-1.amazonaws.com"}, "msg": "On host 
'54.88.146.77' the cert /etc/origin/master/master.etcd-client.crt with CommonName CN=ec2-54-209-71-248.compute-1.amazonaws.com expires in 363 days (Jul 19 21:06:09 2017 GMT)"}
failed: [54.88.146.77] (item={'cert': u'/etc/origin/master/master.kubelet-client.crt', 'expiry': u'Jul 19 21:13:58 2018 GMT', 'days_remaining': 728, 'subject': u'O=system:node-admins, CN=system:openshift-node-admi
n'}) => {"failed": true, "item": {"cert": "/etc/origin/master/master.kubelet-client.crt", "days_remaining": 728, "expiry": "Jul 19 21:13:58 2018 GMT", "subject": "O=system:node-admins, CN=system:openshift-node-adm
in"}, "msg": "On host '54.88.146.77' the cert /etc/origin/master/master.kubelet-client.crt with CommonName O=system:node-admins, CN=system:openshift-node-admin expires in 728 days (Jul 19 21:13:58 2018 GMT)"}
failed: [54.88.146.77] (item={'cert': u'/etc/origin/generated-configs/node-ec2-107-23-232-25.compute-1.amazonaws.com/server.crt', 'expiry': u'Jul 19 21:25:52 2018 GMT', 'days_remaining': 728, 'subject': u'CN=107.2
3.232.25'}) => {"failed": true, "item": {"cert": "/etc/origin/generated-configs/node-ec2-107-23-232-25.compute-1.amazonaws.com/server.crt", "days_remaining": 728, "expiry": "Jul 19 21:25:52 2018 GMT", "subject": "
CN=107.23.232.25"}, "msg": "On host '54.88.146.77' the cert /etc/origin/generated-configs/node-ec2-107-23-232-25.compute-1.amazonaws.com/server.crt with CommonName CN=107.23.232.25 expires in 728 days (Jul 19 21:2
5:52 2018 GMT)"}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @playbooks/adhoc/check-cert-expiry.retry

PLAY RECAP *********************************************************************
54.204.157.7               : ok=6    changed=0    unreachable=0    failed=1
54.88.146.77               : ok=5    changed=0    unreachable=0    failed=1
```

tbielawa commented:
Totally a work in progress. It's a minimum viable product, but it still hides a lot of useful information, such as where extra copies of certificate files are.

detiber commented:
A few notes:
- if you can make this work under the playbooks/common/openshift-cluster and with a wrapper playbook in playbooks/byo/openshift-cluster I think that would be a better approach than adhoc, I want to push adhoc out to the openshift-ansible-contrib repo over time.
- I would suggest that instead of searching for crt files under /etc/origin, you should probably parse the files used in the config files instead, other certificates that are used by pods should be checked by inspecting the secret attached rather than the file on disk, eventually these things should be using the service serving certificates provided by Kube/OpenShift which will take care of rotating out certificates.

tbielawa commented:
New commit responds to suggestion

> if you can make this work under the playbooks/common/openshift-cluster and with a wrapper playbook in playbooks/byo/openshift-cluster I think that would be a better approach than adhoc, I want to push adhoc out to the openshift-ansible-contrib repo over time.

Considering how to approach your other idea and if it's necessary right now. Also, there are certs under the ``/etc/etcd`` directory that could stand to be checked as well.

detiber commented:
@tbielawa etcd is definitely a consideration.  I would propose pulling the paths to those files from /etc/etcd/etcd.conf

tbielawa commented:
Thinking the inventory should enable setting ``cert_expiry_days_warning: 30`` rather than keeping it in ``vars`` in the playbook.

tbielawa commented:
Rebased and squashed onto the latest master. I think these requirements have been met now

> if you can make this work under the playbooks/common/openshift-cluster and with a wrapper playbook in playbooks/byo/openshift-cluster I think that would be a better approach than adhoc

Going to work on switching from reading on the filesystem to reading from OS resources directly whenever possible.

sdodson commented:
Moving out of 3.3 because we added the cert re-deploy.

tbielawa commented:
This is being handled in https://github.com/openshift/openshift-ansible/pull/2549 instead.
